README.md
Codever Browser Extension
---
Browser extension to easily save bookmarks and code snippets from the web to [www.codever.land](https://www.codever.land),
 the Bookmarks and Code Snippets Manager for Developers & Co.
 
## Install
This browser extension is available for:

| [![Chrome](assets/img/chrome-logo-48.png)](https://chrome.google.com/webstore/detail/codever/diofdblfhjbpgackifolmboaiccmebjb) 
|:---:|
| [Chrome](https://chrome.google.com/webstore/detail/codever/diofdblfhjbpgackifolmboaiccmebjb) 

> If you can't use Browser Extensions, or you have a tight security blocking pop-up windows from extensions 
> (**Firefox blocks new windows from pages by default**) you can use [our bookmarklets](https://www.codever.land/howto/bookmarklets)
> which offer the same functionality 

## How to use  

**Right click** OR **click the extension icon** to save as bookmark or snippet to [Codever.land](https://www.codever.land)
- if you make a selection on the web page you will be asked to save as new **snippet** or **bookmark**
- when bookmarking youtube videos and stackoverflow questions the **tags** are auto-completed

### Save bookmark

![Save bookmark demo](assets/img/gif/codever-save-bookmarks-800x454.gif)

### Save snippet

![Save snippet demo](assets/img/gif/codever-save-snippet-800x454.gif)


## Testing locally

Check out the git repository - `git clone https://github.com/codeverland/codever-browser-extension.git`

## Chrome/Brave
Go to [chrome://extensions/](chrome://extensions/), click **Load unpacked** and select the `codever-browser-extension`
where you have checked it out:

![Install locally](assets/img/chrome-install-locally-and-reload-extension.png)

> Click "Reload" on the extension when you do modifications 

### Firefox

#### Use [web-ext](https://github.com/mozilla/web-ext)
The easiest way is to use [web-ext](https://github.com/mozilla/web-ext)
 You can install it globally for example via
```
npm install --global web-ext
```
and then run the following command in the root directory of the project

```
web-ext run
```

This installs **Codever** as a temporary add-on, and it watches for changes in the source code
and **redeploys automatically**.

#### Manual deployment
Go to [about:debugging#/runtime/this-firefox](about:debugging#/runtime/this-firefox), click **Load Temporary Add-on...**
 and select the `manifest.json` file from the `save-code-to-bookmarks.dev-browser-extension` folder where you have checked it out:

![Install locally on Firefox](assets/img/chrome-install-locally-and-reload-extension.png)

### Test the extension against the [`localhost`](https://github.com/codeverland/codever) version of Codever.land
Change the following line in [launch-codever-dialog.js](launch-codever-dialog.js):
```
open('https://www.codever.land/personal/new?url=' + encodeURIComponent(l) + '&desc=' + encodeURIComponent(d) + '&title=' + encodeURIComponent(t) + '&popup=true', 'Codever.land', features);
```
to
```
open('http://localhost:4200/personal/new?url=' + encodeURIComponent(l) + '&desc=' + encodeURIComponent(d) + '&title=' + encodeURIComponent(t) + '&popup=true', 'Codever.land', features);
```

and Reload the extension 

## Publish browser extension to official stores

First of all zip the file either with `web-ext`
```
web-ext build -i 'resources' 'assets' 'README.md' 'CHANGELOG.md'

## with overwrite 
web-ext build --overwrite-dest -i 'resources' 'assets' 'README.md' 'CHANGELOG.md' 
```

or with the _standard_ `zip` command:
```shell
zip -r bookmarks.browser.extension.zip * -x *.idea* *.git* '*resources/*' '*assets/*' "*README.md*" "*CHANGELOG.md*" '*web-ext-artifacts/*'
```

### Publish to Google Chrome Webstore
Go to [Chrome Webstore Dashboard](https://chrome.google.com/webstore/developer/dashboard) where
you upload the .zip file. Wait for a couple of business days for an approval.


### Publish to [Firefox Webstore](https://extensionworkshop.com/documentation/publish/)

Go to [Add-ons page](https://addons.mozilla.org/en-US/developers/addons) and submit the new addon


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.0.0]

Migrate [www.bookmarks.dev](https://www.bookmarks.dev) to [www.codever.land](https://www.codever.land)

## [2.1.0] - 2020-06-27

### Added
- more picture to [docs](resources)
- CHANGELOG

### Changed
- [README.md](./README.md) file
- popup window width, height and positioning to be the same as the bookmarklet


popup.js
/*chrome.runtime.onMessage.addListener(
    function(request, sender, sendResponse) {
        console.log(sender.tab ?
            "from a content script:" + sender.tab.url :
            "from the extension");
        if (request.greeting == "hello"){
            document.getElementById('myIframe').src = "https://www.bookmarks.dev/bookmarklets";
            sendResponse({farewell: "goodbye"});
        }

    });*/

/**
 * When the popup loads, inject a content script into the active tab,
 * and add a click handler.
 * If we couldn't inject the script, handle the error.
 */
/*browser.tabs.executeScript({file: "/popup.js"})
    .then(changeIframe);

function changeIframe() {
    document.getElementById('myIframe').src = "https://www.bookmarks.dev/bookmarklets";
}*/
/*chrome.tabs.query({ active: true, lastFocusedWindow: true }, function (tabs) {
    let url = tabs[0].url;
    document.getElementById('myIframe').src = url;
    //alert(window.getSelection().toString());
});*/

var browser = browser || chrome

window.onload = function() {
    console.log("onload" + Date());
/*    browser.runtime.sendMessage({greeting: "hello"}, function(response) {
        alert(response.farewell);
    });*/
    browser.tabs.query({active: true, currentWindow: true}, function(tabs) {
        browser.tabs.sendMessage(tabs[0].id, {greeting: "helloo"}, function(response) {
            console.log(response);
            //alert(JSON.stringify(response));
            const iframe = document.createElement('iframe');
            //iframe.style.display = "none";
            iframe.src = response.url;
            iframe.width = '' + Math.min(800, parseInt(response.width));
            iframe.height = '' + Math.min(600, parseInt(response.height));
            //alert(iframe.width + ' ' + iframe.height);
            //iframe.src = 'https://www.bookmarks.dev/personal/new';
            const content = document.getElementById('content');
            //content.setAttribute("style","display:block;width:1200px");
            //content.style.width='1200px';*/
            //content.setAttribute("style", 'width:' + response.width + 'px')
            //content.setAttribute("style", 'height:' + response.height + 'px')

            content.appendChild(iframe);
        });
    });
}


launch-codever-dialog.js
{
    const l = location.href;
    let d = '';
    if ( document.getSelection() && document.getSelection().toString().trim() !== '' ) {
        d = document.getSelection();
    }

    const t = document.title;
    let w, h;
    if ( innerWidth > 1400 ) {
        w = 930;
    } else {
        w = (innerWidth * 65) / 100;
    }

    if ( innerHeight > 1400 ) {
        h = 1150;
    } else {
        h = (innerHeight * 90) / 100;
    }
    const features = `toolbar=no,width=${w},height=${h},top=${innerHeight - h}, left=${innerWidth - w}`;
    let url = `https://www.codever.land/my-bookmarks/new?url=${encodeURIComponent(l)}&desc=${encodeURIComponent(d)}&title=${encodeURIComponent(t)}&popup=true`;
    if ( d != '' ) {
        url = `https://www.codever.land/new-entry?url=${encodeURIComponent(l)}&selection=${encodeURIComponent(d)}&title=${encodeURIComponent(t)}&popup=true`;
    }
    open(url, 'Codever', features);
}


content.js
var browser = browser || chrome
browser.runtime.onMessage.addListener(
    function(request, sender, sendResponse) {
        const l = location.href;
        let d = '';
        if ( document.getSelection() && document.getSelection().toString().trim() !== '' ) {
            d = document.getSelection();
        } else {
            d = '';
        }
        const t = document.title;
        if (request.greeting == "helloo") {
            //alert('received helloooo');
            let w, h;
            if ( innerWidth > 1400 ) {
                w = 930;
            } else {
                w = (innerWidth * 65) / 100;
            }

            if ( innerHeight > 1400 ) {
                h = 1150;
            } else {
                h = (innerHeight * 90) / 100;
            }
            const features = `toolbar=no,width=${w},height=${h},top=${innerHeight - h}, left=${innerWidth - w}`;
            //sendResponse({farewell: "goodbye"});
            const url = 'https://www.bookmarks.dev/personal/new?url=' + encodeURIComponent(l) + '&desc=' + encodeURIComponent(d) + '&title=' + encodeURIComponent(t) + '&popupExt=true';
            sendResponse({url: url, width: w, height: h});
            //open(url, 'Bookmarks.dev', features);
        }

        if (request.action=== 'new_tab'){
            const url = 'https://www.bookmarks.dev/personal/new?url=' + encodeURIComponent(l) + '&desc=' + encodeURIComponent(d) + '&title=' + encodeURIComponent(t) + '&popup=true';
            sendResponse({url: url, width: w, height: h});
        }

    });